frame/staking/src/lib.rs

```diff
@@ -1472,18 +1746,173 @@ decl_module! {
+		///
+		/// 1. `min { support.total }` for each support of a winner. This value should be maximized.
+		/// 2. `sum { support.total }` for each support of a winner. This value should be minimized.
+		/// 3. `sum { support.total^2 }` for each support of a winner. This value should be
+		///    minimized (to ensure less variance)
+		///
+		/// # <weight>
+		/// E: number of edges. m: size of winner committee. n: number of nominators. d: edge degree
+		/// (16 for now) v: number of on-chain validator candidates.
+		///
+		/// NOTE: given a solution which is reduced, we can enable a new check the ensure `|E| < n +
+		/// m`. We don't do this _yet_, but our offchain worker code executes it nonetheless.
+		///
+		/// major steps (all done in `check_and_replace_solution`):
+		///
+		/// - Storage: O(1) read `ElectionStatus`.
+		/// - Storage: O(1) read `PhragmenScore`.
+		/// - Storage: O(1) read `ValidatorCount`.
+		/// - Storage: O(1) length read from `SnapshotValidators`.
+		///
+		/// - Storage: O(v) reads of `AccountId` to fetch `snapshot_validators`.
+		/// - Memory: O(m) iterations to map winner index to validator id.
+		/// - Storage: O(n) reads `AccountId` to fetch `snapshot_nominators`.
+		/// - Memory: O(n + m) reads to map index to `AccountId` for un-compact.
+		///
+		/// - Storage: O(e) accountid reads from `Nomination` to read correct nominations.
+		/// - Storage: O(e) calls into `slashable_balance_of_extended` to convert ratio to staked.
+		///
+		/// - Memory: build_support_map. O(e).
+		/// - Memory: evaluate_support: O(E).
+		///
+		/// - Storage: O(e) writes to `QueuedElected`.
+		/// - Storage: O(1) write to `QueuedScore`
+		///
+		/// The weight of this call is 1/10th of the blocks total weight.
+		/// # </weight>
+		#[weight = SimpleDispatchInfo::FixedNormal(100_000_000)]
+		pub fn submit_election_solution(
+			origin,
+			winners: Vec<ValidatorIndex>,
+			compact_assignments: CompactAssignments,
+			score: PhragmenScore,
+		) {
+			let _who = ensure_signed(origin)?;
+			Self::check_and_replace_solution(
+				winners,
+				compact_assignments,
+				ElectionCompute::Signed,
+				score,
+			)?
+		}
+
+		/// Unsigned version of `submit_election_solution`. A signature and claimed validator index
+		/// must be provided. For the call to be validated, the signature must contain the signed
+		/// payload of the first three elements, and match the public key at the claimed validator
+		/// index (using the staking key).
+		#[weight = SimpleDispatchInfo::FixedNormal(100_000_000)]
+		pub fn submit_election_solution_unsigned(
+			origin,
+			winners: Vec<ValidatorIndex>,
+			compact_assignments: CompactAssignments,
+			score: PhragmenScore,
+			// already used and checked in ValidateUnsigned.
+			_validator_index: u32,
+			_signature: <T::KeyType as RuntimeAppPublic>::Signature,
+		) {
+			ensure_none(origin)?;
+			Self::check_and_replace_solution(
+				winners,
+				compact_assignments,
+				ElectionCompute::Unsigned,
+				score,
```

Comment:
Do we need this second function? Can't we just omit any check on the origin, and then a single function would work as both signed and unsigned extrinsic.

Reply:
I don't get the idea. ELI5 or pseudo-code?

Reply:
You have two dispatchable functions which are the same, but one is `ensure_signed` and the other is `ensure_none`. I wonder if this can be combined into one dispatchable function where you do not check the origin for being signed or none.

Reply:
The one for validators has extra fields that are checked in `ValidateUnsigned` (`signature` and `index`).